### PR TITLE
First design of API for micro-problems to be coupled with Micro Manager

### PR DESCRIPTION
**Pull request by IshaanDesai**
This PR defines an API for the micro-problem codes which enable direct coupling to a macro-problem via the Micro Manager and preCICE. A standalone single-physics code which simulates a micro-problem needs to be adapted to the API stated here.

### CODE REVIEW

**Comment by Scriptkiddi** [comment]
You could ship an abstract MicroSimulation class which has to be extended by the implementation. Thus ensuring that every method that the MicroManager needs is implemented.

**Comment by Scriptkiddi** [comment]
The participant name is hardcoded in the micro manager is this intented? in my opinion this should be configurable

**Comment by Scriptkiddi** [comment]
The mesh bounds currently only work for a 2d case, febio as an example is using a 3d mesh so this should be taken into account

**Comment by Scriptkiddi** [comment]
The documentation is missing the file structure that is to be returned by the solve method

**Comment by Scriptkiddi** [comment]
line 110 is unused

**Comment by Scriptkiddi** [comment]
What do you think of switching the write_data list returned by the solve method to a dict indexed with the variable names. 
This should have the advantage that the MicroSimulation can return variables in any order and do not depend on the order in the micro-simulation.json. Because I would guess that that could be a common mistake when setting up a coupling later on.
like this
```
# Configure data written to preCICE
writeDataNames = config.get_write_data_name()
writeDataIDs = {}
for name in writeDataNames:
    writeDataIDs.update({name: interface.get_data_id(name, writeMeshID)})

# Configure data read from preCICE
readDataNames = config.get_read_data_name()
readDataIDs = {}
for name in readDataNames:
    readDataIDs.update({name: interface.get_data_id(name, readMeshID)})

# initialize preCICE
precice_dt = interface.initialize()
dt = min(precice_dt, dt)

# Get macro mesh from preCICE
macroVertexIDs, macroVertexCoords = interface.get_mesh_vertices_and_ids(writeMeshID)
nv, _ = macroVertexCoords.shape

micro_sims = []
for v in range(nv):
    micro_sims.append(MicroSimulation())

writeData = []
list_apap_ext_ = [0] * len(macroVertexIDs)
list_d_vapap_ext__d_apap_ext = [0] * len(macroVertexIDs)
# Initialize coupling data
if interface.is_action_required(precice.action_write_initial_data()):
    interface.write_block_scalar_data(apap_ext_write_data_id, macroVertexIDs, list_apap_ext_)
    interface.write_block_scalar_data(d_vapap_ext__d_apap_ext_data_id, macroVertexIDs, list_d_vapap_ext__d_apap_ext)
    interface.mark_action_fulfilled(precice.action_write_initial_data())

# Initialize coupling data
interface.initialize_data()

t, n = 0, 0
t_checkpoint, n_checkpoint = 0, 0

while interface.is_coupling_ongoing():
    # Write checkpoint
    if interface.is_action_required(precice.action_write_iteration_checkpoint()):
        for v in range(nv):
            micro_sims[v].save_checkpoint()
        t_checkpoint = t
        n_checkpoint = n
        interface.mark_action_fulfilled(precice.action_write_iteration_checkpoint())


    readData = {}
    for name, data_id in readDataIDs.items():
        readData.update({name: interface.read_block_scalar_data(data_id, macroVertexIDs)})

    # Converts the Dicts with list to a list of dicts
    # This switches the "addressing" from attribute lists to a per node basis
    # {attributeName: [List of Values for each vertex], ..} to [{dict for each vertex containing all values},..]
    readDataList = [dict(zip(readData,t)) for t in zip(*readData.values())]
    writeDataList = []
    for i in range(0,len(micro_sims)):
        print(readDataList[i])
        writeDataList.append(micro_sims[i].solve(dt, readDataList[i]))

    # Converts the List of dicts to a dict with lists
    # This reverts back to addressing by attribute lists
    # [{dict containing all output values addressed by attribute name}] to {attrbuteName: [List of values for each Vertex]}
    #
    writeData = {k: [dic[k] for dic in writeDataList] for k in writeDataList[0]}
    for name, data_id in writeDataIDs.items():
        interface.write_block_scalar_data(data_id, macroVertexIDs, writeData.get(name))

    precice_dt = interface.advance(dt)
    dt = min(precice_dt, dt)

    t += dt
    n += 1

    # Restore checkpoint
    if interface.is_action_required(precice.action_read_iteration_checkpoint()):
        for v in range(nv):
            micro_sims[v].revert_to_checkpoint()

        n = n_checkpoint
        t = t_checkpoint
        interface.mark_action_fulfilled(precice.action_read_iteration_checkpoint())
```

**Comment by IshaanDesai** [comment]
> line 110 is unused

At the time of your comments, Line 110 referred to [this line](https://github.com/IshaanDesai/coupled-heat-conduction/pull/7/commits/e743209dd821851512d53db8740a7b196eca821f#diff-e826855c9d8c99b47977d7cb4631f921c494141d0a27273a9e38e707a29df712R110) and the purpose is to determine if the micro-simulation has a method called `initialize`. If it does not have an initialize method, the following serial initialization and writing of initial data to preCICE will be skipped. Does this make sense to you?

**Comment by Scriptkiddi** [comment]
> > line 110 is unused
> 
> At the time of your comments, Line 110 referred to [this line](https://github.com/IshaanDesai/coupled-heat-conduction/pull/7/commits/e743209dd821851512d53db8740a7b196eca821f#diff-e826855c9d8c99b47977d7cb4631f921c494141d0a27273a9e38e707a29df712R110) and the purpose is to determine if the micro-simulation has a method called `initialize`. If it does not have an initialize method, the following serial initialization and writing of initial data to preCICE will be skipped. Does this make sense to you?

I meant this line `k, phi = [], []` line 100 in your pr i think. I probably took the line number from my file merged with your changes sorry for that

File: micro-manager.py [hunk 1]
```diff
@@ -119,42 +120,43 @@ def write_block_matrix_data(m, solver_interface, data_ids, vertex_ids):
...
 
-        k_i = micro_sims[i].solve_heat_cell_problem()
-        k.append(k_i)
+    micro_sims_input = [dict(zip(read_data, t)) for t in zip(*read_data.values())]
+    micro_sims_output = []
+    for i in range(nms):
+        micro_sims_output.append(micro_sims[i].solve(micro_sims_input[i], dt))
 
-        i += 1
+    write_data = {k: [d[k] for d in micro_sims_output] for k in micro_sims_output[0]}
```
Comment (IshaanDesai) [inline micro-manager.py]: @Scriptkiddi I really like these one liners that you suggested, but this specific one is causing some problems. The end result here is a dictionary `write_data` but the values of the dictionary end up being several lists embedded in a list. So for example:

```
write_data["porosity"] = [[0.5], [0.73], [0.78], ...]
```
where the individual lists are values acquired from each micro-simulation. What we need here is one list which is appended with all the values. I tried to get it to work that way but could not get it done. Any ideas?

Comment (Scriptkiddi) [reply micro-manager.py]: We should enforce that scalars and vectors are always wrapped in a list when returned by the micro simulation
than we can use `{k: functools.reduce(operator.iconcat, [dic[k] for dic in a], []) for k in a[0]}`  to flatten the lists for the write_data list

**Pull request merged**